# Linear regression in diabetes dataset

Let's explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [1]:
# Your code here:
import sklearn
from sklearn.datasets import load_diabetes
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
diabetes = (sklearn.datasets.load_diabetes())
diabetes

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990842, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06832974, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286377, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04687948,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452837, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00421986,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
# Your code here:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

In [5]:
# Enter your answer here:
print(f"1) {len(diabetes.keys())}. They are attributes that describe a person's health.")
print("2) diabetes['data'] seems to derive from diabetes['target'], probably the standard deviation from the mean.")
print(f"3) There are {diabetes['target'].shape[0]} records.")

1) 6. They are attributes that describe a person's health.
2) diabetes['data'] seems to derive from diabetes['target'], probably the standard deviation from the mean.
3) There are 442 records.


In [3]:
diabetes['data'], diabetes['target']

(array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990842, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06832974, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286377, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04687948,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452837, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00421986,  0.00306441]]),
 array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59., 104., 182.,
   

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [7]:
# Your code here:


## Buliding a regression model

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [8]:
# Your code here:
from sklearn import linear_model
from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [9]:
# Your code here:
X = diabetes['data']
y = diabetes['target']


diabetes_model = LinearRegression().fit(X, y)
diabetes_model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [10]:
# Your code here:
diabetes_data_train = diabetes['data'][-20:]
diabetes_target_train = diabetes['target'][-20:]


In [11]:
diabetes_data_test = diabetes['data']
diabetes_target_test = diabetes['target']

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [12]:
# Your code here:
killbot = diabetes_model.fit(diabetes_data_train, diabetes_target_train)
print(f'Intercept: {killbot.intercept_}')
print(f'\nCoefficient:\n{killbot.coef_}')

Intercept: 155.69899758090367

Coefficient:
[-3.88886831e+00  2.04648785e+02 -6.42891630e+01 -2.62796691e+02
  1.40037268e+04 -1.17983078e+04 -5.89215807e+03 -1.13694765e+03
 -2.73659711e+03 -3.93879743e+02]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [13]:
# Your code here:
y_diabetes_data_test = killbot.predict(diabetes_data_test)
y_diabetes_data_test[:10]

array([158.8814902 ,  94.34540513, 125.21594995, 148.07503081,
        75.11931221, 118.2753285 , 132.18220485,  74.38163962,
       112.28369699, 354.2283092 ])

#### Print your `diabetes_target_test` and compare with the prediction. 

In [14]:
# Your code here:
for i in range(len(y_diabetes_data_test)):
	print("X= %s, Predicted= %s" % (diabetes_data_test[i], y_diabetes_data_test[i]))

 -0.00290283 -0.00259226 -0.01190068 -0.03835666], Predicted= 89.32572678335076
X= [ 0.0090156  -0.04464164 -0.02021751 -0.0538708   0.03145391  0.02060651
  0.05600338 -0.03949338 -0.01090444 -0.0010777 ], Predicted= 104.52155700994527
X= [ 0.01628068  0.05068012  0.01427248  0.00121513  0.00118295 -0.02135538
 -0.03235593  0.03430886  0.07496834  0.04034337], Predicted= 363.88371873027836
X= [ 0.01991321 -0.04464164 -0.03422907  0.05515344  0.06722868  0.0741549
 -0.00658447  0.03283281  0.02472532  0.06933812], Predicted= 107.23527815053899
X= [ 0.08893144 -0.04464164  0.00672779  0.02531523  0.03007796  0.00870687
  0.06336665 -0.03949338  0.00943641  0.03205916], Predicted= 90.69360642241196
X= [ 0.01991321 -0.04464164  0.00457217  0.04597245 -0.01808039 -0.05454912
  0.06336665 -0.03949338  0.02866072  0.06105391], Predicted= 93.55955701805118
X= [-0.02367725 -0.04464164  0.03043966 -0.00567061  0.08236416  0.09200436
 -0.01762938  0.07120998  0.03304707  0.00306441], Predicted= 

#### Is `diabetes_target_test` exactly the same as the model prediction?

Your explanation here:


No it is not. The reason is simple: the data was not accurate enough. Of 442 entries, only 20 whre selected. Nontheless, with only 20, the accuracy of this model is quite impressive.

In [17]:
r2_score(y_diabetes_data_test, diabetes_data_test)

NameError: name 'r2_score' is not defined